## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [ ]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [89]:
from google.colab import drive
drive.mount('/content/drive')
import os
movies = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Week2/Questions/4_Movies.csv')
movies.head(20)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


,color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,...,num_user_for_reviews,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes
0,Color,James Cameron,723.0,178.0,0.0,855.0,Joel David Moore,1000.0,760505847.0,Action|Adventure|Fantasy|Sci-Fi,...,3054.0,English,USA,PG-13,237000000.0,2009.0,936.0,7.9,1.78,33000
1,Color,Gore Verbinski,302.0,169.0,563.0,1000.0,Orlando Bloom,40000.0,309404152.0,Action|Adventure|Fantasy,...,1238.0,English,USA,PG-13,300000000.0,2007.0,5000.0,7.1,2.35,0
2,Color,Sam Mendes,602.0,148.0,0.0,161.0,Rory Kinnear,11000.0,200074175.0,Action|Adventure|Thriller,...,994.0,English,UK,PG-13,245000000.0,2015.0,393.0,6.8,2.35,85000
3,Color,Christopher Nolan,813.0,164.0,22000.0,23000.0,Christian Bale,27000.0,448130642.0,Action|Thriller,...,2701.0,English,USA,PG-13,250000000.0,2012.0,23000.0,8.5,2.35,164000
4,Color,Andrew Stanton,462.0,132.0,475.0,530.0,Samantha Morton,640.0,73058679.0,Action|Adventure|Sci-Fi,...,738.0,English,USA,PG-13,263700000.0,2012.0,632.0,6.6,2.35,24000
5,Color,Sam Raimi,392.0,156.0,0.0,4000.0,James Franco,24000.0,336530303.0,Action|Adventure|Romance,...,1902.0,English,USA,PG-13,258000000.0,2007.0,11000.0,6.2,2.35,0
6,Color,Nathan Greno,324.0,100.0,15.0,284.0,Donna Murphy,799.0,200807262.0,Adventure|Animation|Comedy|Family|Fantasy|Musi...,...,387.0,English,USA,PG,260000000.0,2010.0,553.0,7.8,1.85,29000
7,Color,Joss Whedon,635.0,141.0,0.0,19000.0,Robert Downey Jr.,26000.0,458991599.0,Action|Adventure|Sci-Fi,...,1117.0,English,USA,PG-13,250000000.0,2015.0,21000.0,7.5,2.35,118000
8,Color,David Yates,375.0,153.0,282.0,10000.0,Daniel Radcliffe,25000.0,301956980.0,Adventure|Family|Fantasy|Mystery,...,973.0,English,UK,PG,250000000.0,2009.0,11000.0,7.5,2.35,10000
9,Color,Zack Snyder,673.0,183.0,0.0,2000.0,Lauren Cohan,15000.0,330249062.0,Action|Adventure|Sci-Fi,...,3018.0,English,USA,PG-13,250000000.0,2016.0,4000.0,6.9,2.35,197000


**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [ ]:
movies.shape
movies.columns
movies.dtypes
movies.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  <font color='red'>(3821, 26)</font>
-  <font color='red'>(3879, 28)</font>
-  <font color='green'>(3853, 28)</font>
-  <font color='red'>(3866, 26)</font>

In [ ]:
movies.shape


(3853, 28)

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='green'>12</font>

In [ ]:
print(movies.isnull().any().sum())

12


### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc. So many of the columns in this dataframe are not required. So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [68]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations


#### <font color='red'>Question 3: What is the count of columns in the new dataframe? </font>
-  <font color='red'>10</font>
-  <font color='green'>13</font>
-  <font color='red'>15</font>
-  <font color='red'>17</font>

In [69]:
list_of_unnecessary_columns = ['color','director_facebook_likes','actor_1_facebook_likes','actor_2_facebook_likes','actor_3_facebook_likes','actor_2_name','cast_total_facebook_likes','actor_3_name','duration', 'facenumber_in_poster','content_rating', 'country', 'movie_imdb_link', 'aspect_ratio', 'plot_keywords'    ]
movies = movies.drop(list_of_unnecessary_columns, axis=1)
print(movies.shape)
print(movies.columns)


(3853, 13)
Index(['director_name', 'num_critic_for_reviews', 'gross', 'genres',
       'actor_1_name', 'movie_title', 'num_voted_users',
       'num_user_for_reviews', 'language', 'budget', 'title_year',
       'imdb_score', 'movie_facebook_likes'],
      dtype='object')


**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'.

In [70]:
movies.isnull().mean() * 100

,0
director_name,0.000000
num_critic_for_reviews,0.025954
gross,0.000000
genres,0.000000
actor_1_name,0.000000
movie_title,0.000000
num_voted_users,0.000000
num_user_for_reviews,0.000000
language,0.103815
budget,0.000000


#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='green'>language</font>
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

In [ ]:
## Write your code here


**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [76]:
# Write your code for filling the NaN values in the 'language' column here

movies['language'] = movies['language'].fillna('English')
movies[movies['language'] == 'English'].shape[0]
movies['language'].value_counts(dropna=False)

,count
language,
English,3675
French,37
Spanish,26
Mandarin,14
German,13
Japanese,12
Hindi,10
Cantonese,8
Italian,7


#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>  <font color='green'>3675?</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [90]:
movies['budget'] = movies['budget'] / 1000000
movies['gross'] = movies['gross'] / 1000000
movies[['budget', 'gross']].head()

,budget,gross
0,237.0,760.505847
1,300.0,309.404152
2,245.0,200.074175
3,250.0,448.130642
4,263.7,73.058679


**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [91]:
movies['profit'] = movies['gross'] - movies['budget']
movies_sorted = movies.sort_values(by='profit', ascending = False)
top10 = movies_sorted.head(10)
top10

,color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,...,language,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes,profit
0,Color,James Cameron,723.0,178.0,0.0,855.0,Joel David Moore,1000.0,760.505847,Action|Adventure|Fantasy|Sci-Fi,...,English,USA,PG-13,237.0,2009.0,936.0,7.9,1.78,33000,523.505847
28,Color,Colin Trevorrow,644.0,124.0,365.0,1000.0,Judy Greer,3000.0,652.177271,Action|Adventure|Sci-Fi|Thriller,...,English,USA,PG-13,150.0,2015.0,2000.0,7.0,2.00,150000,502.177271
25,Color,James Cameron,315.0,194.0,0.0,794.0,Kate Winslet,29000.0,658.672302,Drama|Romance,...,English,USA,PG-13,200.0,1997.0,14000.0,7.7,2.35,26000,458.672302
2704,Color,George Lucas,282.0,125.0,0.0,504.0,Peter Cushing,11000.0,460.935665,Action|Adventure|Fantasy|Sci-Fi,...,English,USA,PG,11.0,1977.0,1000.0,8.7,2.35,33000,449.935665
2748,Color,Steven Spielberg,215.0,120.0,14000.0,548.0,Dee Wallace,861.0,434.949459,Family|Sci-Fi,...,English,USA,PG,10.5,1982.0,725.0,7.9,1.85,34000,424.449459
16,Color,Joss Whedon,703.0,173.0,0.0,19000.0,Robert Downey Jr.,26000.0,623.279547,Action|Adventure|Sci-Fi,...,English,USA,PG-13,220.0,2012.0,21000.0,8.1,1.85,123000,403.279547
482,Color,Roger Allers,186.0,73.0,28.0,847.0,Nathan Lane,2000.0,422.783777,Adventure|Animation|Drama|Family|Musical,...,English,USA,G,45.0,1994.0,886.0,8.5,1.66,17000,377.783777
230,Color,George Lucas,320.0,136.0,0.0,1000.0,Liam Neeson,20000.0,474.544677,Action|Adventure|Fantasy|Sci-Fi,...,English,USA,PG,115.0,1999.0,14000.0,6.5,2.35,13000,359.544677
64,Color,Christopher Nolan,645.0,152.0,22000.0,11000.0,Heath Ledger,23000.0,533.316061,Action|Crime|Drama|Thriller,...,English,USA,PG-13,185.0,2008.0,13000.0,9.0,2.35,37000,348.316061
419,Color,Gary Ross,673.0,142.0,378.0,575.0,Josh Hutcherson,34000.0,407.999255,Adventure|Drama|Sci-Fi|Thriller,...,English,USA,PG-13,78.0,2012.0,14000.0,7.3,2.35,140000,329.999255


**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='green'>E.T. the Extra-Terrestrial</font>
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

In [95]:
top10.iloc[4]['movie_title']


'E.T. the Extra-Terrestrial\xa0'

**Subtask 3.3: Find IMDb Top 200**

Create a new dataframe `IMDb_Top_200` and store the top 200 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`). Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000.

Also add a `Rank` column containing the values 1 to 200 indicating the ranks of the corresponding films.

In [102]:
# Write your code for extracting the top 200 movies as per the IMDb score here. Make sure that you store it in a new dataframe
# and name that dataframe as 'IMDb_Top_200'
IMDb_Top_200 = movies[movies['num_voted_users'] > 25000]
sorted_movies = IMDb_Top_200.sort_values(by='imdb_score', ascending=False)
IMDb_Top_200 = sorted_movies.head(200)
IMDb_Top_200 = IMDb_Top_200.reset_index(drop=True)
IMDb_Top_200['Rank'] = IMDb_Top_200.index + 1
IMDb_Top_200.head(10)




,color,director_name,num_critic_for_reviews,duration,director_facebook_likes,actor_3_facebook_likes,actor_2_name,actor_1_facebook_likes,gross,genres,...,country,content_rating,budget,title_year,actor_2_facebook_likes,imdb_score,aspect_ratio,movie_facebook_likes,profit,Rank
0,Color,Frank Darabont,199.0,142.0,0.0,461.0,Jeffrey DeMunn,11000.0,28.341469,Crime|Drama,...,USA,R,25.0,1994.0,745.0,9.3,1.85,108000,3.341469,1
1,Color,Francis Ford Coppola,208.0,175.0,0.0,3000.0,Marlon Brando,14000.0,134.821952,Crime|Drama,...,USA,R,6.0,1972.0,10000.0,9.2,1.85,43000,128.821952,2
2,Color,Christopher Nolan,645.0,152.0,22000.0,11000.0,Heath Ledger,23000.0,533.316061,Action|Crime|Drama|Thriller,...,USA,PG-13,185.0,2008.0,13000.0,9.0,2.35,37000,348.316061,3
3,Color,Francis Ford Coppola,149.0,220.0,0.0,3000.0,Al Pacino,22000.0,57.300000,Crime|Drama,...,USA,R,13.0,1974.0,14000.0,9.0,1.85,14000,44.300000,4
4,Color,Sergio Leone,181.0,142.0,0.0,24.0,Luigi Pistilli,16000.0,6.100000,Western,...,Italy,Approved,1.2,1966.0,34.0,8.9,2.35,20000,4.900000,5
5,Color,Peter Jackson,328.0,192.0,0.0,416.0,Billy Boyd,5000.0,377.019252,Action|Adventure|Drama|Fantasy,...,USA,PG-13,94.0,2003.0,857.0,8.9,2.35,16000,283.019252,6
6,Color,Quentin Tarantino,215.0,178.0,16000.0,857.0,Eric Stoltz,13000.0,107.930000,Crime|Drama,...,USA,R,8.0,1994.0,902.0,8.9,2.35,45000,99.930000,7
7,Black and White,Steven Spielberg,174.0,185.0,14000.0,212.0,Embeth Davidtz,14000.0,96.067179,Biography|Drama|History,...,USA,R,22.0,1993.0,795.0,8.9,1.85,41000,74.067179,8
8,Color,Peter Jackson,297.0,171.0,0.0,857.0,Orlando Bloom,16000.0,313.837577,Action|Adventure|Drama|Fantasy,...,New Zealand,PG-13,93.0,2001.0,5000.0,8.8,2.35,21000,220.837577,9
9,Color,Christopher Nolan,642.0,148.0,22000.0,23000.0,Tom Hardy,29000.0,292.568851,Action|Adventure|Sci-Fi|Thriller,...,USA,PG-13,160.0,2010.0,27000.0,8.8,2.35,175000,132.568851,10


#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='green'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_200*? </font>

In [107]:
b1 = ((IMDb_Top_200['imdb_score'] >= 7.5) & (IMDb_Top_200['imdb_score'] < 8)).sum()
b2 = ((IMDb_Top_200['imdb_score'] >= 8) & (IMDb_Top_200['imdb_score'] < 8.5)).sum()
b3 = ((IMDb_Top_200['imdb_score'] >= 8.5) & (IMDb_Top_200['imdb_score'] < 9)).sum()
b4 = ((IMDb_Top_200['imdb_score'] >= 9) & (IMDb_Top_200['imdb_score'] < 9.5)).sum()
b5 = ((IMDb_Top_200['imdb_score'] >= 9.5) & (IMDb_Top_200['imdb_score'] <= 10)).sum()

counts = [b1, b2, b3, b4, b5]
labels = ['7.5-8', '8-8.5', '8.5-9', '9-9.5', '9.5-10']

labels[counts.index(max(counts))]


'8-8.5'

**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [109]:
Meryl_Streep = movies[movies['actor_1_name'] == 'Meryl Streep']

In [110]:
Leo_Caprio = movies[movies['actor_1_name'] == 'Leonardo DiCaprio']

In [111]:
Brad_Pitt = movies[movies['actor_1_name'] == 'Brad Pitt']

In [113]:
Combined = pd.concat([Meryl_Streep, Leo_Caprio, Brad_Pitt])

In [114]:
grouped = Combined.groupby('actor_1_name')



In [119]:
grouped['num_critic_for_reviews'].idxmax()
grouped['num_user_for_reviews'].idxmax()



,num_user_for_reviews
actor_1_name,
Brad Pitt,646
Leonardo DiCaprio,94
Meryl Streep,1784


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='green'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [131]:
grouped['num_user_for_reviews'].mean().idxmax()


'Leonardo DiCaprio'

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='green'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

In [132]:
grouped['num_critic_for_reviews'].mean().idxmax()


'Leonardo DiCaprio'

#### <font color='red'>Question 10: Which movie starting with the letter A, has the 3rd highest profit?</font>
-  <font color='red'>Austin Powers: The Spy Who Shagged Me</font>
-  <font color='red'>Avatar</font>
-  <font color='green'>Avengers: Age of Ultron</font>

In [133]:
movies_A = movies[movies['movie_title'].str.startswith('A')]
movies_A_sorted = movies_A.sort_values(by='profit', ascending=False)
movies_A_sorted.iloc[2]['movie_title']

'Avengers: Age of Ultron\xa0'

#### <font color='red'>Question 11: Among the movies starting with the letter A, who has the highest imdb score?</font>
-  <font color='green'>American History X</font>
-  <font color='red'>Apocalypse Now</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [134]:
movies_A_sorted = movies_A.sort_values(by='imdb_score', ascending=False)
movies_A_sorted.iloc[0]['movie_title']

'American History X\xa0'

#### <font color='red'>Question 12: Among the movies that '*Joss Whedon*' has directed, which of them made the highest profit?</font>
-  <font color='red'>Serenity</font>
-  <font color='green'>The Avengers</font>
-  <font color='red'>Avengers: Age of Ultron</font>

In [135]:
movies_Joss = movies[movies['director_name']== 'Joss Whedon']
movies_Joss_sorted = movies_Joss.sort_values(by='profit', ascending=False)
movies_Joss_sorted.iloc[0]['movie_title']


'The Avengers\xa0'

#### <font color='red'>Question 13: Among the movies that were released in the year '*2007*' & '*2008*', which of them had the 2nd highest imdb score?</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='green'>WALL·E</font>
-  <font color='red'>No End in Sight</font>
-  <font color='red'>Into the Wild</font>

In [139]:
movies_07_08 = movies[movies['title_year'].isin([2007,2008])]
movies_07_08_sorted = movies_07_08.sort_values(by='imdb_score', ascending=False)
movies_07_08_sorted.iloc[1]['movie_title']

'WALL·E\xa0'